File: ios/Runner/LivenessPlugin.swift
import Foundation
import Flutter
import AVFoundation

/// Flutter plugin for Liveness Detection on iOS.
/// Matches Android LivenessPlugin functionality.
class LivenessPlugin: NSObject, FlutterPlugin {
    
    private let methodChannel: FlutterMethodChannel
    private let eventChannel: FlutterEventChannel
    private var eventSink: FlutterEventSink?
    
    private var textureRegistry: FlutterTextureRegistry?
    private var textureId: Int64 = -1
    
    // Pipeline
    private var pipeline: LivenessPipeline?
    
    // Components (for pipeline creation)
    private var faceDetector: FaceDetectorProtocol?
    private var landmarkExtractor: LandmarkExtractorProtocol?
    private var screenManager: ScreenManagerProtocol?
    
    // Camera
    private var captureSession: AVCaptureSession?
    private var videoOutput: AVCaptureVideoDataOutput?
    private let videoQueue = DispatchQueue(label: "com.diajarkoding.livenessdetection.videoQueue")
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "com.diajarkoding.livenessdetection/method",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "com.diajarkoding.livenessdetection/events",
            binaryMessenger: registrar.messenger()
        )
        
        let instance = LivenessPlugin(
            methodChannel: methodChannel,
            eventChannel: eventChannel,
            textureRegistry: registrar.textures()
        )
        
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }
    
    init(methodChannel: FlutterMethodChannel, eventChannel: FlutterEventChannel, textureRegistry: FlutterTextureRegistry?) {
        self.methodChannel = methodChannel
        self.eventChannel = eventChannel
        self.textureRegistry = textureRegistry
        super.init()
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            initialize(result: result)
        case "startCamera":
            startCamera(result: result)
        case "stopCamera":
            stopCamera(result: result)
        case "startVerification":
            startVerification(result: result)
        case "reset":
            reset(result: result)
        case "dispose":
            dispose(result: result)
        case "warmUp":
            warmUp(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    /// Initialize vision components without starting camera.
    /// Called during instruction screen for warm-up.
    /// If already initialized, just reset the pipeline state.
    private func initialize(result: @escaping FlutterResult) {
        // If pipeline already exists, just reset it and return success
        if let existingPipeline = pipeline {
            print("[LivenessPlugin] Pipeline already exists, resetting instead of re-initializing")
            existingPipeline.reset()
            result(["success": true])
            return
        }
        
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            // Create components
            self.faceDetector = MLKitFaceDetector()
            self.landmarkExtractor = MediaPipeLandmarker()
            self.screenManager = IOSScreenManager()
            
            // Create pipeline
            self.pipeline = LivenessPipeline(
                faceDetector: self.faceDetector!,
                landmarkExtractor: self.landmarkExtractor!,
                screenManager: self.screenManager!
            )
            
            // Set up state change callback
            self.pipeline?.onStateChanged = { [weak self] state in
                self?.sendStateUpdate(state)
            }
            
            // Initialize pipeline (includes warm-up)
            self.pipeline?.initialize { success in
                DispatchQueue.main.async {
                    if success {
                        result(["success": true])
                        print("[LivenessPlugin] Initialization complete")
                    } else {
                        result(FlutterError(code: "INIT_ERROR", message: "Failed to initialize pipeline", details: nil))
                    }
                }
            }
        }
    }
    
    /// Warm up MediaPipe engine before camera starts.
    private func warmUp(result: @escaping FlutterResult) {
        landmarkExtractor?.warmUp {
            result(["success": true])
        }
    }
    
    /// Start AVCaptureSession for camera preview.
    private func startCamera(result: @escaping FlutterResult) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            let session = AVCaptureSession()
            session.sessionPreset = .photo
            
            guard let device = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .front),
                  let input = try? AVCaptureDeviceInput(device: device) else {
                result(FlutterError(code: "CAMERA_ERROR", message: "Could not access camera", details: nil))
                return
            }
            
            if session.canAddInput(input) {
                session.addInput(input)
            }
            
            let output = AVCaptureVideoDataOutput()
            output.videoSettings = [kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA]
            output.setSampleBufferDelegate(self, queue: self.videoQueue)
            
            if session.canAddOutput(output) {
                session.addOutput(output)
            }
            
            self.captureSession = session
            self.videoOutput = output
            
            DispatchQueue.global(qos: .userInitiated).async {
                session.startRunning()
            }
            
            // Start gating phase now that camera is ready
            self.pipeline?.startGating()
            
            // For now, return a placeholder texture ID
            // In production, you'd use FlutterTexture for zero-copy rendering
            result([
                "textureId": 0,
                "width": 480,
                "height": 640
            ])
            
            print("[LivenessPlugin] Camera started")
        }
    }
    
    /// Stop camera and release resources.
    private func stopCamera(result: @escaping FlutterResult) {
        captureSession?.stopRunning()
        captureSession = nil
        videoOutput = nil
        result(nil)
        print("[LivenessPlugin] Camera stopped")
    }
    
    /// Start verification process.
    private func startVerification(result: @escaping FlutterResult) {
        pipeline?.reset()
        result(nil)
        print("[LivenessPlugin] Verification started")
    }
    
    /// Reset pipeline state.
    private func reset(result: @escaping FlutterResult) {
        pipeline?.reset()
        result(nil)
        print("[LivenessPlugin] Pipeline reset")
    }
    
    /// Dispose all resources.
    private func dispose(result: @escaping FlutterResult) {
        captureSession?.stopRunning()
        captureSession = nil
        videoOutput = nil
        
        pipeline?.dispose()
        pipeline = nil
        
        faceDetector = nil
        landmarkExtractor = nil
        screenManager = nil
        
        result(nil)
        print("[LivenessPlugin] Plugin disposed")
    }
    
    /// Send state update to Flutter via EventChannel.
    private func sendStateUpdate(_ state: LivenessState) {
        var data: [String: Any]
        
        switch state {
        case .idle:
            data = ["state": "idle"]
            
        case .initializing:
            data = ["state": "initializing"]
            
        case .gating(let message):
            data = [
                "state": "gating",
                "message": message
            ]
            
        case .gatingFailed(let reason, let message):
            data = [
                "state": "gating_failed",
                "reason": reason.rawValue,
                "message": message
            ]
            
        case .challenge(let type, let progress, let instruction):
            data = [
                "state": "challenge",
                "type": type.rawValue,
                "progress": progress,
                "instruction": instruction
            ]
            
        case .processing(let message):
            data = [
                "state": "processing",
                "message": message
            ]
            
        case .success(let completedChallenges):
            data = [
                "state": "success",
                "challenges": completedChallenges.map { $0.rawValue }
            ]
            
        case .failed(let reason, let canRetry):
            data = [
                "state": "failed",
                "reason": reason,
                "canRetry": canRetry
            ]
        }
        
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(data)
        }
    }
}

// MARK: - FlutterStreamHandler
extension LivenessPlugin: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        print("[LivenessPlugin] Event channel listening")
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        print("[LivenessPlugin] Event channel cancelled")
        return nil
    }
}

// MARK: - AVCaptureVideoDataOutputSampleBufferDelegate
extension LivenessPlugin: AVCaptureVideoDataOutputSampleBufferDelegate {
    func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
        // Process frame through the pipeline
        pipeline?.processFrame(sampleBuffer: sampleBuffer)
    }
}

File: android/app/src/main/kotlin/com/diajarkoding/livenessdetection/bridge/LivenessPlugin.kt
package com.diajarkoding.livenessdetection.bridge

import android.app.Activity
import android.content.Context
import android.util.Log
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import com.diajarkoding.livenessdetection.core.*
import com.diajarkoding.livenessdetection.impl.AndroidScreenManager
import com.diajarkoding.livenessdetection.impl.mediapipe.MediaPipeLandmarker
import com.diajarkoding.livenessdetection.impl.mlkit.MLKitFaceDetector
import com.diajarkoding.livenessdetection.logic.*
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.collectLatest
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Flutter plugin for Liveness Detection.
 * Bridges Flutter UI with native vision pipeline.
 */
class LivenessPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, 
    ActivityAware, EventChannel.StreamHandler {
    
    companion object {
        private const val TAG = "LivenessPlugin"
        private const val METHOD_CHANNEL = "com.diajarkoding.livenessdetection/method"
        private const val EVENT_CHANNEL = "com.diajarkoding.livenessdetection/events"
    }
    
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    
    private var context: Context? = null
    private var activity: Activity? = null
    private var textureRegistry: TextureRegistry? = null
    private var textureEntry: TextureRegistry.SurfaceTextureEntry? = null
    
    private var cameraProvider: ProcessCameraProvider? = null
    private var cameraExecutor: ExecutorService? = null
    private var imageAnalysis: ImageAnalysis? = null
    private var preview: Preview? = null
    
    private var pipeline: LivenessPipeline? = null
    private var scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private var stateCollectorJob: Job? = null
    
    // Components
    private var faceDetector: IFaceDetector? = null
    private var landmarkExtractor: ILandmarkExtractor? = null
    private var screenManager: IScreenManager? = null
    
    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        textureRegistry = binding.textureRegistry
        
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)
        
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)
        
        cameraExecutor = Executors.newSingleThreadExecutor()
        
        Log.d(TAG, "Plugin attached to engine")
    }
    
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        dispose()
        Log.d(TAG, "Plugin detached from engine")
    }
    
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        Log.d(TAG, "Plugin attached to activity")
    }
    
    override fun onDetachedFromActivity() {
        activity = null
        Log.d(TAG, "Plugin detached from activity")
    }
    
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }
    
    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }
    
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        Log.d(TAG, "Event channel listening")
        
        // Start collecting state updates
        pipeline?.let { p ->
            scope.launch {
                p.state.collectLatest { state ->
                    sendStateUpdate(state)
                }
            }
        }
    }
    
    override fun onCancel(arguments: Any?) {
        eventSink = null
        Log.d(TAG, "Event channel cancelled")
    }
    
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> initialize(result)
            "startCamera" -> startCamera(result)
            "stopCamera" -> stopCamera(result)
            "startVerification" -> startVerification(result)
            "reset" -> reset(result)
            "dispose" -> {
                dispose()
                result.success(null)
            }
            "warmUp" -> warmUp(result)
            else -> result.notImplemented()
        }
    }
    
    /**
     * Initialize vision components without starting camera.
     * Called during instruction screen for warm-up.
     * If already initialized, just reset the pipeline state.
     */
    private fun initialize(result: MethodChannel.Result) {
        val ctx = context ?: run {
            result.error("NO_CONTEXT", "Context not available", null)
            return
        }
        
        val act = activity ?: run {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }
        
        // Ensure scope is active (recreate if cancelled)
        if (!scope.isActive) {
            Log.d(TAG, "Recreating cancelled scope")
            scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
        }
        
        // If pipeline already exists, just reset it and return success
        pipeline?.let { existingPipeline ->
            Log.d(TAG, "Pipeline already exists, resetting instead of re-initializing")
            existingPipeline.reset()
            result.success(mapOf("success" to true))
            return
        }
        
        scope.launch {
            try {
                // Create components
                faceDetector = MLKitFaceDetector()
                landmarkExtractor = MediaPipeLandmarker(ctx)
                screenManager = AndroidScreenManager(act)
                
                // Run security checks
                val securityResult = SecurityGuard.performSecurityCheck(ctx)
                Log.d(TAG, "Security check: safe=${securityResult.isSafe}, score=${securityResult.riskScore}, factors=${securityResult.riskFactors}")
                
                // Create pipeline
                pipeline = LivenessPipeline(
                    faceDetector!!,                    
                    landmarkExtractor!!,
                    screenManager!!,
                    securityResult
                )
                
                // Initialize pipeline (includes warm-up)
                pipeline?.initialize()
                
                // Cancel previous state collector if exists
                stateCollectorJob?.cancel()
                
                // Start collecting state updates
                stateCollectorJob = pipeline?.let { p ->
                    scope.launch {
                        p.state.collectLatest { state ->
                            sendStateUpdate(state)
                        }
                    }
                }
                
                result.success(mapOf("success" to true))
                Log.d(TAG, "Initialization complete")
            } catch (e: Exception) {
                Log.e(TAG, "Initialization failed", e)
                result.error("INIT_ERROR", e.message, null)
            }
        }
    }
    
    /**
     * Warm up MediaPipe engine before camera starts.
     */
    private fun warmUp(result: MethodChannel.Result) {
        scope.launch {
            try {
                landmarkExtractor?.warmUp()
                result.success(mapOf("success" to true))
            } catch (e: Exception) {
                result.error("WARMUP_ERROR", e.message, null)
            }
        }
    }
    
    /**
     * Start CameraX with SurfaceTexture for Flutter rendering.
     */
    @androidx.camera.core.ExperimentalGetImage
    private fun startCamera(result: MethodChannel.Result) {
        val ctx = context ?: run {
            result.error("NO_CONTEXT", "Context not available", null)
            return
        }
        
        val act = activity as? LifecycleOwner ?: run {
            result.error("NO_LIFECYCLE", "Activity must be LifecycleOwner", null)
            return
        }
        
        // Ensure camera executor exists
        if (cameraExecutor == null || cameraExecutor!!.isShutdown) {
            cameraExecutor = Executors.newSingleThreadExecutor()
        }
        
        val cameraProviderFuture = ProcessCameraProvider.getInstance(ctx)
        
        cameraProviderFuture.addListener({
            try {
                cameraProvider = cameraProviderFuture.get()
                
                // Create texture for Flutter
                textureEntry = textureRegistry?.createSurfaceTexture()
                val surfaceTexture = textureEntry?.surfaceTexture()
                
                if (surfaceTexture == null) {
                    result.error("TEXTURE_ERROR", "Failed to create surface texture", null)
                    return@addListener
                }
                
                // Configure preview with 4:3 ratio
                preview = Preview.Builder()
                    .setTargetAspectRatio(AspectRatio.RATIO_4_3)
                    .build()
                
                preview?.setSurfaceProvider { request ->
                    surfaceTexture.setDefaultBufferSize(
                        request.resolution.width,
                        request.resolution.height
                    )
                    request.provideSurface(
                        android.view.Surface(surfaceTexture),
                        cameraExecutor!!
                    ) { }
                }
                
                // Configure image analysis with target rotation
                val displayRotation = activity?.windowManager?.defaultDisplay?.rotation ?: android.view.Surface.ROTATION_0
                imageAnalysis = ImageAnalysis.Builder()
                    .setTargetAspectRatio(AspectRatio.RATIO_4_3)
                    .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                    .setTargetRotation(displayRotation)
                    .build()
                
                imageAnalysis?.setAnalyzer(cameraExecutor!!) { imageProxy ->
                    pipeline?.processFrame(imageProxy)
                }
                
                // Use front camera
                val cameraSelector = CameraSelector.Builder()
                    .requireLensFacing(CameraSelector.LENS_FACING_FRONT)
                    .build()
                
                // Bind use cases
                cameraProvider?.unbindAll()
                cameraProvider?.bindToLifecycle(
                    act,
                    cameraSelector,
                    preview,
                    imageAnalysis
                )
                
                val textureId = textureEntry?.id() ?: -1L
                
                // Start gating phase now that camera is ready
                pipeline?.startGating()
                
                result.success(mapOf(
                    "textureId" to textureId,
                    "width" to 480,
                    "height" to 640
                ))
                
                Log.d(TAG, "Camera started with texture ID: $textureId")
            } catch (e: Exception) {
                Log.e(TAG, "Failed to start camera", e)
                result.error("CAMERA_ERROR", e.message, null)
            }
        }, ContextCompat.getMainExecutor(ctx))
    }
    
    /**
     * Stop camera and release resources.
     */
    private fun stopCamera(result: MethodChannel.Result) {
        cameraProvider?.unbindAll()
        textureEntry?.release()
        textureEntry = null
        result.success(null)
        Log.d(TAG, "Camera stopped")
    }
    
    /**
     * Start verification process.
     */
    private fun startVerification(result: MethodChannel.Result) {
        pipeline?.reset()
        result.success(null)
        Log.d(TAG, "Verification started")
    }
    
    /**
     * Reset pipeline state.
     */
    private fun reset(result: MethodChannel.Result) {
        pipeline?.reset()
        result.success(null)
        Log.d(TAG, "Pipeline reset")
    }
    
    /**
     * Dispose all resources.
     */
    private fun dispose() {
        // Cancel state collector first
        stateCollectorJob?.cancel()
        stateCollectorJob = null
        
        // Cancel scope
        scope.cancel()
        
        cameraProvider?.unbindAll()
        cameraProvider = null
        
        textureEntry?.release()
        textureEntry = null
        
        cameraExecutor?.shutdown()
        cameraExecutor = null
        
        pipeline?.dispose()
        pipeline = null
        
        faceDetector = null
        landmarkExtractor = null
        screenManager = null
        
        Log.d(TAG, "Plugin disposed")
    }
    
    /**
     * Send state update to Flutter via EventChannel.
     */
    private fun sendStateUpdate(state: LivenessState) {
        val data = when (state) {
            is LivenessState.Idle -> mapOf(
                "state" to "idle"
            )
            is LivenessState.Initializing -> mapOf(
                "state" to "initializing"
            )
            is LivenessState.Gating -> mapOf(
                "state" to "gating",
                "message" to state.message
            )
            is LivenessState.GatingFailed -> mapOf(
                "state" to "gating_failed",
                "reason" to state.reason.name,
                "message" to state.message
            )
            is LivenessState.Challenge -> mapOf(
                "state" to "challenge",
                "type" to state.type.name,
                "progress" to state.progress,
                "instruction" to state.instruction
            )
            is LivenessState.Processing -> mapOf(
                "state" to "processing",
                "message" to state.message
            )
            is LivenessState.Success -> mapOf(
                "state" to "success",
                "challenges" to state.completedChallenges.map { it.name }
            )
            is LivenessState.Failed -> mapOf(
                "state" to "failed",
                "reason" to state.reason,
                "canRetry" to state.canRetry
            )
        }
        
        activity?.runOnUiThread {
            eventSink?.success(data)
        }
    }
}
